Create ConstantSPOSet for Unit Testing

rcclay opened the pull request:
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

This adds an SPOSet object that returns constant values for value, gradient, and laplacians.  These constant values are set at construction.  Basically, just set and keep Nelec x Norb matrices.  evaluateValue, evaluateVGL, and evaluate_notranspose just return slices of these matrices.  Similar spirit as FakeSPOSet, but with the ability to use arbitrarily sized and valued matrices.  

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Intel Xeon, GNU 9.2.0, OpenMPI  1.10.1
## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

prckent commented:
Since this is a new testing thing, please aim for really high coverage.

rcclay commented:
@prckent Fair point.  Will do.

markdewing commented on src/QMCWaveFunctions/tests/test_ConstantSPOSet.cpp:
```diff
@@ -0,0 +1,193 @@
+
+namespace qmcplusplus
+{
+//Ray:  Figure out how to template me on value type.
```
For reference, this test uses TEMPLATE_TEST_CASE: https://github.com/QMCPACK/qmcpack/blob/19b99c950d1bb518332c69bda4adf823adab97d1/src/Numerics/tests/test_RotationMatrix3D.cpp#L20-L23

rcclay replied:
Thank you!   I'll try to make more efficient use of template_test_case in the future.  For now, real vs. complex test cases are covered by letting the imaginary parts be zero.

ye-luo commented:
We have a class called "ConstantOrbital". The added one further caused confusion.

> Similar spirit as FakeSPOSet, but with the ability to use arbitrarily sized and valued matrices.

Can we fuse FakeSPOSet and this one and avoid increasing maintenance complexity.

rcclay commented:
@ye-luo For first point, it's definitely confusing, but I'd argue it's the fault of ConstantOrbital's name.  It's really a WaveFunctionComponent.  I think it was overlooked in renaming Orbitals to WaveFunctionComponents.  

For second point, I think the two could probably be collapsed into one class, but @markdewing is better to comment on it than me.  In the mean time, I need to look to see what all would be affected.

ye-luo commented:
> @ye-luo For first point, it's definitely confusing, but I'd argue it's the fault of ConstantOrbital's name. It's really a WaveFunctionComponent. I think it was overlooked in renaming Orbitals to WaveFunctionComponents.

agreed. ConstantOrbital was bad in the first place.

> For second point, I think the two could probably be collapsed into one class, but @markdewing is better to comment on it than me. In the mean time, I need to look to see what all would be affected.

Please discuss with @markdewing I think FakeSPO needs some extension to accommodate your needs. it is too rigid right now. The outcome of your discussion will be a good foundation for writing future tests involving SPOSet.

prckent commented:
Besides ConstantOrbital, we also have a LinearOrbital that is also a WaveFuctionComponent. The latter does get invoked by the test_dmc test.  Can we rename these?

rcclay commented:
For this PR, I'm going to:

- [ ] Replace FakeSPO instances with ConstantSPOSet.  
- [ ] Change API away from setting value, gradient, etc. matrices in inititalization towards getter/setter functions.

markdewing commented:
I have one concern about the "Constant" part of the name.  With ConstantOrbital and LinearOrbital, the first part of the name refers to mathematical properties, and the value and derivatives are consistent.  The "Constant" in ConstantSPO means "fixed" or "evaluated at a single point".   I'm okay with either meaning, but I'm wondering if the name "Constant" with two different uses would be confusing?  Or if the differences are clear enough from reading the code?  (Since ConstantOrbital needs to be renamed eventually for the "Orbital" part, the first word could be changed as well?)

rcclay commented:
ReferenceSPOSet?  ContainerSPOSet?

ye-luo commented:
If all the use cases of previous FakeSPOSet are covered. I would just claim it FakeSPOSet.

PDoakORNL commented on src/QMCWaveFunctions/ConstantSPOSet.h:
```diff
@@ -0,0 +1,133 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
```
Mock classes should go in the testing directories not module roots.  See `src/Estimators/tests/FakeOperatorEstimator.h`  Generally they should not be header only. Making the tests build slower benefits nobody.  The targets they are built as part of shouldn't be linked into the application code.

If the class is with testing code and only linked to a test program or sub target of a test program it may not be necessary to call it Fake... or Mock...

rcclay replied:
My intended use case here is for Estimator/Hamiltonian evaluation, secondarily as a test of WaveFunctionComponents, so they would be linked to stuff in both of those directories.  Specifically, want to get rid of some of the SPOSet gymnastics done in things like test_soecp.cpp (and soon to be in magnetization density).  What do you recommend in this case?  Same location and naming convention?

PDoakORNL replied:
Right now we do share tests classes across modules, all the minimal pools for instance.  At the moment the easier thing to do in that case is to leave the test class header only. Include by full path to the file from the project root if crossing modules.
ex:
```
#include "Particle/tests/MinimalParticlePool.h"
```

PDoakORNL replied:
The only one that seems to make a target that gets linked by other tests is container_testing from the Containers module. But I think its overkill for this class.

PDoakORNL commented on src/QMCWaveFunctions/tests/test_ConstantSPOSet.cpp:
```diff
@@ -0,0 +1,193 @@
+  const int nelec = 2;
+  const int norb  = 2;
+
+  using RealType    = QMCTraits::RealType;
```
I'd suggest using src/QMCWaveFunctions/WaveFunctionTypes.hpp to define consistent types.  QMCTraits bakes in the multiexecutable design and is just bad. OrbitalSetTraits obfuscates the handling of precision.

example from `DiracDeterminantBatched`
```
  using WFT           = WaveFunctionTypes<VALUE, FP_VALUE>;
  using Value         = typename WFT::Value;
  using FullPrecValue = typename WFT::FullPrecValue;
  using PsiValue      = typename WFT::PsiValue;
  using LogValue      = typename WFT::LogValue;
  using Grad          = typename WFT::Grad;
  using Hess          = typename WFT::Hess;
  using Real          = typename WFT::Real;
```
Vector types gain no value from the typedefs at all since they can just be written
```
Vector<Value> row0;
Vector<Grad> grow0;
Vector<Value> lrow0l;
```

rcclay replied:
WaveFunctionTypes is the file I need, thank you.  I'll make those changes

rcclay replied:
Done

PDoakORNL commented on src/QMCWaveFunctions/tests/test_ConstantSPOSet.cpp:
```diff
@@ -0,0 +1,193 @@
+  CHECK(psiV[0] == row1[0]);
+  CHECK(psiV[1] == row1[1]);
+  
+  CHECK(psiG[0][0] == grow1[0][0]); 
```
Use `src/Utils/forTesting/checkMatrix.hpp`?  This is painful to look at and it has pretty decent reporting if there is an bad element.

rcclay replied:
Also exact tool I need.  Thank you.  Will fix.

rcclay replied:
Do we have anything like checkMatrix for vectors?

rcclay replied:
ParticleGradient and GradTypes are vectors of vectors, so for now I'm just checking them with loops.  Need to write some checker functions to handle vectors and vectors of vectors.

PDoakORNL reviewed (changes requested):
Class needs to be moved, there is some redundancy with FakeSpoSet but I don't think addressing that is hugely important in this PR, it could just be a good first issue for someone to work on.

rcclay commented:
@ye-luo I took a stab at merging FakeSPO and ConstantSPOSet, and I think this might be better as a separate PR.  FakeSPO returns values for evaluateValue, evaluateVGL, evaluateAlltoOne, etc., through a mixture of returning matrix values set at construction, and on the fly-computation.  That would be fine by itself, but there's also some logic going on between plain evaluate_notranspose like calls  evaluateV* for ratio evaluations using an active particle that will take some disentangling, and maybe some tweaking of the tests.

rcclay commented:
Hi all--
Recent changes:

1. Moved ConstantSPOSet stuff to QMCWaveFunction/tests
2. Split implementation from header. 
3. Replaced matrix comparisons with checkMatrix().  
4. Added documentation to ConstantSPOSet header and setter functions.

ye-luo commented on src/QMCWaveFunctions/tests/ConstantSPOSet.h:
```diff
@@ -0,0 +1,85 @@
+   *  Exists to provide deterministic and known output to objects requiring SPOSet evaluations.      
+   *
+   */
+struct ConstantSPOSet : public SPOSet
```
Class instead of struct.

rcclay replied:
Whoops.  Done.

ye-luo commented on src/QMCWaveFunctions/tests/ConstantSPOSet.h:
```diff
@@ -0,0 +1,85 @@
+
+protected:
+private:
+  int numparticles_;
```
`const int numparticles_;` as this is set at the constructor.
Could you add a comment about this variable? I feel the name can be very confusing and documentation should help.

rcclay replied:
Done.

ye-luo commented on src/QMCWaveFunctions/tests/ConstantSPOSet.h:
```diff
@@ -0,0 +1,85 @@
+                            GradMatrix& dlogdet,
+                            ValueMatrix& d2logdet) override;
+
+protected:
```
Remove it please.

rcclay replied:
Done

ye-luo commented:
@PDoakORNL good to be merged?

PDoakORNL commented:
I'm satisfied

ye-luo commented:
Test this please
